### ISSUE TITLE
No JSON object could be decoded

### ISSUE THREAD

**Issue by rukamesh-kumar** [golastmile/rasa_nlu#205]
**Rasa NLU version**  : `latest version`

**Used backend**  : `mitie`
**Plattform**  :  `Ubuntu-14.04`

**Issue:**
I have tried to train on a file generated using rasa-nlu-trainer, but i am getting below error .

**Full stack trace :** 
`Traceback (most recent call last):
  File "/usr/lib64/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "build/bdist.linux-x86_64/egg/rasa_nlu/train.py", line 72, in <module>
  File "build/bdist.linux-x86_64/egg/rasa_nlu/train.py", line 64, in do_train
  File "build/bdist.linux-x86_64/egg/rasa_nlu/training_data.py", line 35, in __init__
  File "build/bdist.linux-x86_64/egg/rasa_nlu/training_data.py", line 78, in guess_format
  File "/usr/lib64/python2.7/json/__init__.py", line 339, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python2.7/json/decoder.py", line 364, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib64/python2.7/json/decoder.py", line 382, in raw_decode
    raise ValueError("No JSON object could be decoded")
`

I have validated using json validator it is correct json file.
we have attached config.json for mitie and also tried with mitie_sklearn getting  same issue.

**Attachments**
 Data file : [rasa.json.zip](https://github.com/golastmile/rasa_nlu/files/827946/rasa.json.zip)
Config file : [config.json.zip](https://github.com/golastmile/rasa_nlu/files/827956/config.json.zip)

**Comment by tmbo** [comment]
Just tried it out locally with your provided files and I couldn't reproduce the issue. Maybe its some sort of encoding issue? (eg. windows vs linux)

Please make sure the configuration location is correct as well.

**Comment by 2fat2kidnap** [comment]
Having the same here. Pulled directly from the current live Git repo (i.e. not throug `pip`).  
System: `macOS 10.11.6`  
Python: `2.7.10`  
Backend: `mitie`
JSON validated with JSON linting. Pulled from the current web demo of the [data visualiser](https://golastmile.github.io/rasa-nlu-trainer/). Interestingly, running it with `demo-rasa.json` instead produces no errors/runs succesfully.  
Traceback:  
```
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "build/bdist.macosx-10.11-intel/egg/rasa_nlu/train.py", line 72, in <module>
  File "build/bdist.macosx-10.11-intel/egg/rasa_nlu/train.py", line 64, in do_train
  File "build/bdist.macosx-10.11-intel/egg/rasa_nlu/training_data.py", line 35, in __init__
  File "build/bdist.macosx-10.11-intel/egg/rasa_nlu/training_data.py", line 78, in guess_format
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py", line 338, in loads
    return _default_decoder.decode(s)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 366, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 384, in raw_decode
    raise ValueError("No JSON object could be decoded")
```

I'll mention it here if I find what's up.

**Comment by tmbo** [comment]
Ok so @2fat2kidnap just for me to be able to reproduce this:

1. which version of the visualizer did you use, the local one or the hosted one?
2. you upload your data, download it and then it doesn't work anymore?

**Comment by 2fat2kidnap** [comment]
1) I used the hosted one at the link I posted in my previous post. 
* I hit the clear button to remove the example entries
* I then added entries manually with the `Add new example` button
* Once I was done, I downloaded the file with the `Download` button
The file I obtained through downloading throws an encoding error. The file provided in the `rasa` directory as an example works without a hitch.

**Comment by 2fat2kidnap** [comment]
So directly downloading the example file from the [hosted config-file generator](https://golastmile.github.io/rasa-nlu-trainer/), without making any edits to it, also throws the encoding error mentioned above. Is that reproducible for you @tmbo ?

I've tried going through it with a TextDiff tool, comparing the `hosted.json` vs the `demo-rasa.json`, and the only difference I'm noting is with regards to spaces after elements, but I don't think that should affect the JSON parsing? There must be some kind of encoding going wrong somewhere, but I haven't figured out yet where. 

Will report back if I figure out more.

**Comment by tmbo** [comment]
Alright, thanks for that detailed explanation, but i couldnt yet reproduce it. maybe its a browser "feature", which one are you using?

**Comment by 2fat2kidnap** [comment]
Using Chrome `Version 56.0.2924.87 (64-bit)` currently.  
I've dug into a bunch of file encoding information, here's what I'm getting (still on macOS):
```
$ file testData.json 
testData.json: UTF-8 Unicode (with BOM) English text
```
versus
```
$ file demo-rasa.json 
demo-rasa.json: ASCII English text
```

I know Python2.7 tends to act up with respect to encodings, maybe the issue is here somewhere? I can't get `iconv` to work right now:  
```
iconv -f utf-8 -t ascii ayylmao.json -o "ayylmao.utf8.json"
iconv: ayylmao.json:1:0: cannot convert
```
but I'm sure that's because I'm a complete amateur with these command-line tools. If I manage to convert, I'll report back what the results are.

**Comment by 2fat2kidnap** [comment]
Well, this was entertaining, I learned of a new UNIX utility, today: `iconv`

**TL;DR**
`iconv -c -f UTF8 -t ASCII original.json > originalEncoded.json` fixes the issue.

**Somewhat more detailed, but very hypothetical, explanation:**
Looks like Chrome browser (at least, I'm assuming it's tied to Chrome given that it was hard to reproduce) pushes for `UTF-8` encoding when downloading files. On the other hand, it looks like `rasa_nlu.train` expects `ASCII` encoded files. Using: 
`iconv -c -f UTF8 -t ASCII original.json > originalEncoded.json`  
converts the downloaded file to ASCII-readable encoding, allowing Python2.7 to play nicely with it.
Don't forget the `-c` flag, or you'll get a `iconv: could not convert` error (at least, I did).

If you're not sure you have the right encodings installed on your machine (you should though, this stuff is default), just issue `iconv -l`, it'll list all the encodings.

**Comment by tmbo** [comment]
Thanks for your help. The issue is with the BOM (which actually should be avoided for json files, I am not sure why chrome puts them there, might need to have a look at the download code). 

Long story short, I fixed to code to cope with files having BOMs. 60e2625
